<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
        <title>Model the Gene Expression (2): Likelihood Ratio Test - Taotao Tan</title><meta name="Description" content="Taotao&#39;s Blog Posts"><meta property="og:title" content="Model the Gene Expression (2): Likelihood Ratio Test" />
<meta property="og:description" content="In the last post, we used a GLM framework to model the gene expression. $$ y \sim NB(\mu, r) \\ log( \mu )= b_1 x &#43; b_0 $$
Using maximum likelihood estimation, we were able to find a set of parameters $\hat b_0, \hat b_1, \hat r$, that maximizes the likelihood function.
But if you send this model (the estimated parameters) to biologists, they wouldn&rsquo;t be happy. And we all know what is lacking: the p-value!" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://jasontan-code.github.io/posts/gene-exp2/" />
<meta property="og:image" content="https://jasontan-code.github.io/images/avatar.png"/>
<meta property="article:published_time" content="2021-11-23T15:19:26-06:00" />
<meta property="article:modified_time" content="2021-11-23T15:19:26-06:00" />
<meta name="twitter:card" content="summary_large_image"/>
<meta name="twitter:image" content="https://jasontan-code.github.io/images/avatar.png"/>

<meta name="twitter:title" content="Model the Gene Expression (2): Likelihood Ratio Test"/>
<meta name="twitter:description" content="In the last post, we used a GLM framework to model the gene expression. $$ y \sim NB(\mu, r) \\ log( \mu )= b_1 x &#43; b_0 $$
Using maximum likelihood estimation, we were able to find a set of parameters $\hat b_0, \hat b_1, \hat r$, that maximizes the likelihood function.
But if you send this model (the estimated parameters) to biologists, they wouldn&rsquo;t be happy. And we all know what is lacking: the p-value!"/>
<meta name="application-name" content="Taotao&#39;s Blog Posts">
<meta name="apple-mobile-web-app-title" content="Taotao&#39;s Blog Posts"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png"><link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="/site.webmanifest"><link rel="canonical" href="https://jasontan-code.github.io/posts/gene-exp2/" /><link rel="prev" href="https://jasontan-code.github.io/posts/gene-exp1/" /><link rel="next" href="https://jasontan-code.github.io/posts/bayesian-network-1/" /><link rel="stylesheet" href="/lib/normalize/normalize.min.css"><link rel="stylesheet" href="/css/style.min.css"><link rel="stylesheet" href="/lib/fontawesome-free/all.min.css"><link rel="stylesheet" href="/lib/animate/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "Model the Gene Expression (2): Likelihood Ratio Test",
        "inLanguage": "en",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https:\/\/jasontan-code.github.io\/posts\/gene-exp2\/"
        },"genre": "posts","keywords": "Genetics","wordcount":  1485 ,
        "url": "https:\/\/jasontan-code.github.io\/posts\/gene-exp2\/","datePublished": "2021-11-23T15:19:26-06:00","dateModified": "2021-11-23T15:19:26-06:00","publisher": {
            "@type": "Organization",
            "name": ""},"author": {
                "@type": "Person",
                "name": "Taotao"
            },"description": ""
    }
    </script></head>
    <body header-desktop="fixed" header-mobile="auto"><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('auto' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : 'auto' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="/" title="Taotao Tan">Taotao Tan</a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="/posts/"> Posts </a><a class="menu-item" href="/categories/"> Categories </a><a class="menu-item" href="/about/"> About </a><span class="menu-item delimiter"></span><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                    <i class="fas fa-adjust fa-fw"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="/" title="Taotao Tan">Taotao Tan</a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><a class="menu-item" href="/posts/" title="">Posts</a><a class="menu-item" href="/categories/" title="">Categories</a><a class="menu-item" href="/about/" title="">About</a><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                <i class="fas fa-adjust fa-fw"></i>
            </a></div>
    </div>
</header>
<div class="search-dropdown desktop">
    <div id="search-dropdown-desktop"></div>
</div>
<div class="search-dropdown mobile">
    <div id="search-dropdown-mobile"></div>
</div>
<main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">Contents</h2>
            <div class="toc-content" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">Model the Gene Expression (2): Likelihood Ratio Test</h1><div class="post-meta">
            <div class="post-meta-line"><span class="post-author"><a href="/" title="Author" rel=" author" class="author"><i class="fas fa-user-circle fa-fw"></i>Taotao</a></span>&nbsp;<span class="post-category">included in <a href="/categories/genetics/"><i class="far fa-folder fa-fw"></i>Genetics</a></span></div>
            <div class="post-meta-line"><i class="far fa-calendar-alt fa-fw"></i>&nbsp;<time datetime="2021-11-23">2021-11-23</time>&nbsp;<i class="fas fa-pencil-alt fa-fw"></i>&nbsp;1485 words&nbsp;
                <i class="far fa-clock fa-fw"></i>&nbsp;7 minutes&nbsp;</div>
        </div><div class="details toc" id="toc-static"  kept="true">
                <div class="details-summary toc-title">
                    <span>Contents</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a></li>
    <li><a href="#anova">ANOVA</a></li>
    <li><a href="#likelihood-ratio-test">Likelihood Ratio Test</a></li>
    <li><a href="#calculation">Calculation</a>
      <ul>
        <li></li>
      </ul>
    </li>
    <li><a href="#summary">Summary</a></li>
    <li><a href="#reference">Reference:</a></li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><p>In the <a href="https://jasontan-code.github.io/blog/gene-exp1/" target="_blank" rel="noopener noreffer">last post</a>, we used a GLM framework to model the gene expression.
$$
y \sim NB(\mu, r) \\
log( \mu )= b_1 x + b_0
$$</p>
<p>Using <a href="https://jasontan-code.github.io/posts/mle/" target="_blank" rel="noopener noreffer">maximum likelihood estimation</a>, we were able to find a set of parameters $\hat b_0, \hat b_1, \hat r$, that maximizes the likelihood function.</p>
<p>But if you send this model (the estimated parameters) to biologists, they wouldn&rsquo;t be happy. And we all know what is lacking: <strong>the p-value!</strong></p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="introduction">Introduction</h2>
<p>Obviously, it&rsquo;s inappropriate to use a t-test/ANOVA to evaluate a generalized linear model, since the assumptions of linearity and normality fail. Now the question is: do we have a generic method to construct a hypothesis test, under the GLM framework?</p>
<p>The answers is YES, and even better, we have more than one method to get the job done, namely:</p>
<ul>
<li>Likelihood ratio test</li>
<li>Wald test</li>
<li>Score test</li>
</ul>
<p>Wald test and Score test are quite complicated, and myself is still trying to wrap my head around them. Hopefully I will have a chance to discuss them in future blog posts. But in this post, we will focus on the likelihood ratio test. I will offer some intuition, compare likelihood ratio test to ANOVA, and show you the calculation step by step,</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="anova">ANOVA</h2>
<p>I think it might be helpful to go over ANOVA before we dive into likelihood ratio test, since they share a lot of similarities.
 </p>
<p>Recall in ANOVA, we build two models: a simple model, and a fancy model. By adding an additional variable, we want to see if the Sum of Squared Residuals decreases significantly, comparing to the inherent noise.</p>
<p> </p>
<p>Take our type 2 diabetic medication as an example. A simple model could be &ldquo;the medication doesn&rsquo;t affect gene1 expression&rdquo;. A fancy model could be &ldquo;the medication changed gene 1 expression&rdquo;.</p>
<p> </p>
<p>We fit the simple model using <code>lm(data ~ 1)</code>, and fit the fancy model using <code>lm(data ~ expgroup)</code>. After fitting those two models, we can calculate the Sum of Square Residuals (SSR) for each model: $SSR_{\text{simple}}, SSR_{\text{fancy}}$. Notice adding an additional variable could never hurt the model, so we have $SSR_{\text{simple}} &gt; SSR_{\text{fancy}}$, for any variables.</p>
<p> </p>
<p>The next step of ANOVA is to plug $SSR_{\text{simple}}, SSR_{\text{fancy}}$ into the formula</p>
<p>$$
F = \frac{ (SSR_{\text{simple}} - SSR_{\text{fancy}}) /(p_{\text{fancy}} - p_{\text{simple}}) }{SSR_{\text{fancy}}/(n - p_{\text{fancy}} ) }
$$</p>
<p>, where $p_{\text{fancy}}, p_{\text{simple}}$ represents the number of parameters in that model.</p>
<p> </p>
<p>We calculate the $F$, which has a f-distribution with $df_1 = p_{\text{fancy}} - p_{\text{simple}}$,  $df_2 = n - p_{\text{fancy}}$. To find the p value, we calculate the probability of more extreme events, which is $f$ greater than what we have observed (our calculated F) &ndash; this is the area of the upper tail of the F distribution.</p>
<p> </p>
<p>The take home message is that, we are comparing a simple model and a fancy model. We want to determine if adding another variable could significantly improve the fitness (measured in SSR) of the model.</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="likelihood-ratio-test">Likelihood Ratio Test</h2>
<p>Similarly, in likelihood ratio test, we also fit a simple model, and a fancy model. In our medication example, our simple model could be expressed as <code>glm.nb(data ~ 1)</code>, while the fancy model could be expressed as <code>glm.nb(data ~ expgroup)</code>.</p>
<p> </p>
<p>To evaluate the fitness of the model, we use <a href="https://jasontan-code.github.io/blog/mle/" target="_blank" rel="noopener noreffer">maximum likelihood estimation</a> again. For the simple model, we find the maximum of its likelihood function $\mathcal{L}_m(\mu, r | data)$; for the fancy model, we also find its maximum of likelihood function $\mathcal{L}_m(b_0, b_1, r | data)$.</p>
<p>Then we take the logarithm of the likelihoods ratio:</p>
<p> <br>
$$
\begin{aligned}
\lambda &amp; = -2 \cdot ln \Bigg[ \frac{\mathcal{L}_{\text{m}}(\mu, r | data)}{\mathcal{L} _{\text{m}}(b_0, b_1, r | data)} \Bigg] \\
&amp; = -2 \cdot \Bigg[ ln (\mathcal{L}_m (\mu, r | data)) - ln (\mathcal{L}_m (b_0, b_1,  r | data)) \Bigg]
\end{aligned}
$$</p>
<p>It has been proved that $\lambda$ is <strong>asymptotically</strong>*  Chi square distributed, with a degree of freedom of 1 (in a more general case, with a degree of freedom equal to the difference in dimension between the fancy model and simple model).</p>
<p> </p>
<p>To find the p value, we simply calculate the area of the upper tail of the $\chi^2$ distribution.</p>
<p> </p>
<p>I hope you have find some similarities between the Likelihood Ratio Test and ANOVA. The procedures are almost identical:</p>
<ul>
<li>build a simple model and a fancy model</li>
<li>calculate the fitness for each model (RSS vs likelihood)</li>
<li>determine if adding another variable could significantly improve the fitness</li>
</ul>
<p> </p>
<p>(*<strong>Asymptotically:</strong> it means as the sample size gets larger, $\lambda$ tends to be closer to a theoretical chi square distribution. With that being said, when we have a small sample size, the p value calculated with chi square distribution will be a little bit off)</p>
<p>(I should mention that internally <strong>ANOVA is a likelihood ratio test</strong>, but I will skip the discussion for conciseness)</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="calculation">Calculation</h2>
<p>Now let&rsquo;s do the actual calculation</p>
<table>
<thead>
<tr>
<th style="text-align:left">Patient</th>
<th style="text-align:right">Treatment</th>
<th style="text-align:right">Gene1 Expression</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">patient1</td>
<td style="text-align:right">placebo (0)</td>
<td style="text-align:right">1000</td>
</tr>
<tr>
<td style="text-align:left">patient2</td>
<td style="text-align:right">placebo (0)</td>
<td style="text-align:right">400</td>
</tr>
<tr>
<td style="text-align:left">patient3</td>
<td style="text-align:right">placebo (0)</td>
<td style="text-align:right">700</td>
</tr>
<tr>
<td style="text-align:left">patient4</td>
<td style="text-align:right">medication (1)</td>
<td style="text-align:right">2100</td>
</tr>
<tr>
<td style="text-align:left">patient5</td>
<td style="text-align:right">medication (1)</td>
<td style="text-align:right">3200</td>
</tr>
<tr>
<td style="text-align:left">patient6</td>
<td style="text-align:right">medication (1)</td>
<td style="text-align:right">1000</td>
</tr>
</tbody>
</table>
<p> <br>
 </p>
<h4 id="step-1-build-a-simple-model">Step 1: Build a simple model</h4>
<p>This is equivalent to say that all the patients are from the same distribution. We construct the likelihood function as what we have done before</p>
<p>$$
\begin{aligned}
\mathcal{L}(\mu, r | data) = &amp; \prod_{i = 1}^6 P(y_i |\mu, r)
\end{aligned}
$$</p>
<p> <br>
 </p>
<h4 id="step-2-build-a-fancy-model">Step 2: Build a fancy model</h4>
<p>$$
\begin{aligned}
\mathcal{L}(b_0, b_1, r | data) = &amp; \prod_{i = 1}^6 P(y_i |b_0, b_1, r)
\end{aligned}
$$</p>
<p> <br>
 </p>
<h4 id="step-3-find-the-log-maximum-likelihood-for-the-simple-model">Step 3: Find the log maximum likelihood for the simple model</h4>
<p>Again, we are going to use <code>optim()</code> function to help us do the calculation</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="n">data</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1000</span><span class="p">,</span> <span class="m">400</span><span class="p">,</span> <span class="m">700</span><span class="p">,</span> <span class="m">2100</span><span class="p">,</span> <span class="m">3200</span><span class="p">,</span> <span class="m">1000</span><span class="p">)</span>
<span class="n">expgroup</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span> <span class="p">))</span> 
 

<span class="n">neg_loglikelihood_simple</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">params</span><span class="p">){</span>
  <span class="n">mu</span> <span class="o">=</span> <span class="n">params[1]</span> <span class="c1"># intercept</span>
  <span class="n">r</span> <span class="o">=</span> <span class="n">params[2]</span> <span class="c1"># dispersion</span>
  
  <span class="c1"># define the log likelihood of each patient</span>
  <span class="c1"># patient 1-3 received placebo, therefore mu = b0</span>
  <span class="n">loglikelihood1</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[1]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  <span class="n">loglikelihood2</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[2]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  <span class="n">loglikelihood3</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[3]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  
  <span class="c1"># patient 4-6 received medication, therefore mu = b0 + b1</span>
  <span class="n">loglikelihood4</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[4]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  <span class="n">loglikelihood5</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[5]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  <span class="n">loglikelihood6</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[6]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>

  <span class="c1"># summation of log, equivalent to log product.</span>
  <span class="c1"># optim function finds the minimum, we maximize the loglikelihood, which is equivalent to minimize neg_loglikelihood</span>
  <span class="nf">return</span><span class="p">(</span><span class="o">-</span> <span class="p">(</span><span class="n">loglikelihood1</span> <span class="o">+</span> <span class="n">loglikelihood2</span> <span class="o">+</span> <span class="n">loglikelihood3</span> <span class="o">+</span> <span class="n">loglikelihood4</span> <span class="o">+</span> <span class="n">loglikelihood5</span> <span class="o">+</span> <span class="n">loglikelihood6</span><span class="p">))</span>
<span class="p">}</span>

<span class="nf">optim</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">20</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> <span class="n">neg_loglikelihood_simple</span><span class="p">)</span>

</code></pre></div><p>We found
$$
\hat \mu = 1400, \hat r = 2.31 \\
ln (\mathcal{L}_m (\mu, r | data)) = -48.5
$$</p>
<p> <br>
 </p>
<h4 id="step-4-find-the-log-maximum-likelihood-for-the-fancy-model">Step 4: Find the log maximum likelihood for the fancy model</h4>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="n">neg_loglikelihood_fancy</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">params</span><span class="p">){</span>
  <span class="n">b0</span> <span class="o">=</span> <span class="n">params[1]</span> <span class="c1"># intercept</span>
  <span class="n">b1</span> <span class="o">=</span> <span class="n">params[2]</span> <span class="c1"># coefficient</span>
  <span class="n">r</span> <span class="o">=</span> <span class="n">params[3]</span> <span class="c1"># dispersion</span>
  
  <span class="c1"># define the log likelihood of each patient</span>
  <span class="c1"># patient 1-3 received placebo, therefore mu = b0</span>
  <span class="n">loglikelihood1</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[1]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span><span class="p">)))</span>
  <span class="n">loglikelihood2</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[2]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span><span class="p">)))</span>
  <span class="n">loglikelihood3</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[3]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span><span class="p">)))</span>
  
  <span class="c1"># patient 4-6 received medication, therefore mu = b0 + b1</span>
  <span class="n">loglikelihood4</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[4]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span> <span class="o">+</span> <span class="n">b1</span><span class="p">)))</span>
  <span class="n">loglikelihood5</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[5]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span> <span class="o">+</span> <span class="n">b1</span><span class="p">)))</span>
  <span class="n">loglikelihood6</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[6]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span> <span class="o">+</span> <span class="n">b1</span><span class="p">)))</span>

  <span class="c1"># summation of log, equivalent to log product.</span>
  <span class="c1"># optim function finds the minimum, we maximize the loglikelihood, which is equivalent to minimize neg_loglikelihood</span>
  <span class="nf">return</span><span class="p">(</span><span class="o">-</span> <span class="p">(</span><span class="n">loglikelihood1</span> <span class="o">+</span> <span class="n">loglikelihood2</span> <span class="o">+</span> <span class="n">loglikelihood3</span> <span class="o">+</span> <span class="n">loglikelihood4</span> <span class="o">+</span> <span class="n">loglikelihood5</span> <span class="o">+</span> <span class="n">loglikelihood6</span><span class="p">))</span>
<span class="p">}</span>


<span class="nf">optim</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> <span class="n">neg_loglikelihood_fancy</span><span class="p">)</span>

</code></pre></div><p>We found
$$
\hat b_0 = 6.55, \hat b_1 = 1.09, \hat r = 5.92 \\
ln (\mathcal{L}_m (b_0, b_1, r | data)) = -45.4
$$</p>
<p> <br>
 </p>
<h4 id="step-5-calculate-lambda-and-p">Step 5: Calculate $\lambda$ and p</h4>
<p>$$
\begin{aligned}
\lambda  &amp; = -2 \cdot (ln (\mathcal{L}_m (\mu, r | data)) - ln (\mathcal{L}_m (b_0, b_1, r | data))) \\
&amp; = -2 \cdot (−48.5 - -45.4) \\
&amp; = 6.2
\end{aligned}
$$</p>
<p>Plug into $\chi^2$ distribution with $df = 1$, we calculate the p value:</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="nf">pchisq</span><span class="p">(</span><span class="m">6.2</span><span class="p">,</span> <span class="n">df</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">lower.tail</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
</code></pre></div><p>We have
$$
p = 0.013
$$</p>
<p> <br>
 <br>
 <br>
 </p>
<p>You may use the <code>lmtest::lrtest()</code> function to confirm the calculation:</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="nf">library</span><span class="p">(</span><span class="n">MASS</span><span class="p">)</span>
<span class="n">data</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1000</span><span class="p">,</span> <span class="m">400</span><span class="p">,</span> <span class="m">700</span><span class="p">,</span> <span class="m">2100</span><span class="p">,</span> <span class="m">3200</span><span class="p">,</span> <span class="m">1000</span><span class="p">)</span>
<span class="n">expgroup</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span> <span class="p">))</span>

<span class="c1"># build a simple model and a fancy model</span>
<span class="n">model_simple</span> <span class="o">=</span> <span class="nf">glm.nb</span><span class="p">(</span><span class="n">data</span> <span class="o">~</span> <span class="m">1</span><span class="p">)</span>
<span class="n">model_fancy</span> <span class="o">=</span> <span class="nf">glm.nb</span><span class="p">(</span><span class="n">data</span> <span class="o">~</span> <span class="n">expgroup</span><span class="p">)</span>

<span class="c1"># perform test</span>
<span class="nf">lrtest</span><span class="p">(</span><span class="n">model_simple</span><span class="p">,</span> <span class="n">model_fancy</span><span class="p">)</span>
</code></pre></div><p>You should find the results to be the same!</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="summary">Summary</h2>
<p>In this post, we explored the likelihood ratio test, and found $p = 0.013$, which is sufficient to reject the null hypothesis if we set $\alpha = 0.05$.</p>
<p>I should reinforce the interpretation of the p value, regardless what statistical tests we are using:</p>
<p><strong>CORRECT INTERPRETATION:</strong> P value is the probability of observing the events or more extreme events, given the null hypothesis is true.</p>
<p><strong>WRONG INTERPRETATION:</strong> P value is the probability of making a mistake.</p>
<p> <br>
 <br>
 <br>
 <br>
 <br>
 </p>
<h2 id="reference">Reference:</h2>
<p>Wiki likelihood ratio test: <a href="https://en.wikipedia.org/wiki/Likelihood-ratio_test#cite_note-2">https://en.wikipedia.org/wiki/Likelihood-ratio_test#cite_note-2</a></p>
<p>Statquest: <a href="https://www.youtube.com/watch?v=NF5_btOaCig">https://www.youtube.com/watch?v=NF5_btOaCig</a></p>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>Updated on 2021-11-23</span>
            </div>
            <div class="post-info-license"></div>
        </div>
        <div class="post-info-line">
            <div class="post-info-md"></div>
            <div class="post-info-share">
                <span><a href="javascript:void(0);" title="Share on Twitter" data-sharer="twitter" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test" data-via="doubleTaoTan" data-hashtags="Genetics"><i class="fab fa-twitter fa-fw"></i></a><a href="javascript:void(0);" title="Share on Facebook" data-sharer="facebook" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-hashtag="Genetics"><i class="fab fa-facebook-square fa-fw"></i></a><a href="javascript:void(0);" title="Share on WhatsApp" data-sharer="whatsapp" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test" data-web><i class="fab fa-whatsapp fa-fw"></i></a><a href="javascript:void(0);" title="Share on Line" data-sharer="line" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test"><i data-svg-src="/lib/simple-icons/icons/line.min.svg"></i></a><a href="javascript:void(0);" title="Share on 微博" data-sharer="weibo" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test"><i class="fab fa-weibo fa-fw"></i></a><a href="javascript:void(0);" title="Share on Myspace" data-sharer="myspace" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test" data-description=""><i data-svg-src="/lib/simple-icons/icons/myspace.min.svg"></i></a><a href="javascript:void(0);" title="Share on Blogger" data-sharer="blogger" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test" data-description=""><i class="fab fa-blogger fa-fw"></i></a><a href="javascript:void(0);" title="Share on Evernote" data-sharer="evernote" data-url="https://jasontan-code.github.io/posts/gene-exp2/" data-title="Model the Gene Expression (2): Likelihood Ratio Test"><i class="fab fa-evernote fa-fw"></i></a></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"><i class="fas fa-tags fa-fw"></i>&nbsp;<a href="/tags/genetics/">Genetics</a></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">Back</a></span>&nbsp;|&nbsp;<span><a href="/">Home</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="/posts/gene-exp1/" class="prev" rel="prev" title="Model the Gene Expression (1): A GLM framework"><i class="fas fa-angle-left fa-fw"></i>Model the Gene Expression (1): A GLM framework</a>
            <a href="/posts/bayesian-network-1/" class="next" rel="next" title="An Intuitive Explanation of Bayesian Network">An Intuitive Explanation of Bayesian Network<i class="fas fa-angle-right fa-fw"></i></a></div>
</div>
<div id="comments"><div id="fb-root" class="comment"></div>
            <div
                class="fb-comments"
                data-href="https://jasontan-code.github.io/posts/gene-exp2/"
                data-width="100%"
                data-numposts="10"
            ></div><noscript>
                Please enable JavaScript to view the comments powered by <a href="https://developers.facebook.com/docs/plugins/comments/"></a>Facebook</a>.
            </noscript></div></article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">Powered by <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.80.0">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"><i class="far fa-kiss-wink-heart fa-fw"></i> LoveIt</a>
                </div><div class="footer-line"><i class="far fa-copyright fa-fw"></i><span itemprop="copyrightYear">2021</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="/" target="_blank"></a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="Back to Top">
                <i class="fas fa-arrow-up fa-fw"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="View Comments">
                <i class="fas fa-comment fa-fw"></i>
            </a>
        </div><link rel="stylesheet" href="/lib/lightgallery/lightgallery.min.css"><link rel="stylesheet" href="/lib/katex/katex.min.css"><link rel="stylesheet" href="/lib/katex/copy-tex.min.css"><link rel="stylesheet" href="/lib/cookieconsent/cookieconsent.min.css"><script type="text/javascript" src="https://connect.facebook.net/en_US/sdk.js#xfbml=1&amp;version=v5.0&amp;appId=2162149760591137&amp;autoLogAppEvents=1" defer></script><script type="text/javascript" src="/lib/smooth-scroll/smooth-scroll.min.js"></script><script type="text/javascript" src="/lib/lazysizes/lazysizes.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lightgallery.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lg-thumbnail.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lg-zoom.min.js"></script><script type="text/javascript" src="/lib/clipboard/clipboard.min.js"></script><script type="text/javascript" src="/lib/sharer/sharer.min.js"></script><script type="text/javascript" src="/lib/katex/katex.min.js"></script><script type="text/javascript" src="/lib/katex/auto-render.min.js"></script><script type="text/javascript" src="/lib/katex/copy-tex.min.js"></script><script type="text/javascript" src="/lib/katex/mhchem.min.js"></script><script type="text/javascript" src="/lib/cookieconsent/cookieconsent.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"Copy to clipboard","maxShownLines":10},"comment":{},"cookieconsent":{"content":{"dismiss":"Got it!","link":"Learn more","message":"This website uses Cookies to improve your experience."},"enable":true,"palette":{"button":{"background":"#f0f0f0"},"popup":{"background":"#1aa3ff"}},"theme":"edgeless"},"lightGallery":{"actualSize":false,"exThumbImage":"data-thumbnail","hideBarsDelay":2000,"selector":".lightgallery","speed":400,"thumbContHeight":80,"thumbWidth":80,"thumbnail":true},"math":{"delimiters":[{"display":true,"left":"$$","right":"$$"},{"display":true,"left":"\\[","right":"\\]"},{"display":false,"left":"$","right":"$"},{"display":false,"left":"\\(","right":"\\)"}],"strict":false}};</script><script type="text/javascript" src="/js/theme.min.js"></script><script type="text/javascript">
            window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments);}gtag('js', new Date());
            gtag('config', 'G-C4MZRQHWVE', { 'anonymize_ip': true });
        </script><script type="text/javascript" src="https://www.googletagmanager.com/gtag/js?id=G-C4MZRQHWVE" async></script></body>
</html>
