<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
        <title>Model the Gene Expression (1): A GLM framework - Taotao Tan</title><meta name="Description" content="Taotao&#39;s Blog Posts"><meta property="og:title" content="Model the Gene Expression (1): A GLM framework" />
<meta property="og:description" content="Before you start reading this post, please familiarize yourself with MLE and linear model.
 Background In transcriptomic research, we often want to determine if genes are unregulated or down-regulated under a particular perturbation. For example, we have a medication that may cure type 2 diabetes. In our experiment, 6 patients are split into two groups, with 3 patients taking the medication, and 3 patients taking the placebo. The patients&#39; blood samples are then collected to measure the transcriptomic profile (mRNA abundance level for each gene) using NGS technology (RNA seq)." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://jasontan-code.github.io/posts/gene-exp1/" />
<meta property="og:image" content="https://jasontan-code.github.io/images/avatar.png"/>
<meta property="article:published_time" content="2021-11-23T15:18:26-06:00" />
<meta property="article:modified_time" content="2021-11-23T15:18:26-06:00" />
<meta name="twitter:card" content="summary_large_image"/>
<meta name="twitter:image" content="https://jasontan-code.github.io/images/avatar.png"/>

<meta name="twitter:title" content="Model the Gene Expression (1): A GLM framework"/>
<meta name="twitter:description" content="Before you start reading this post, please familiarize yourself with MLE and linear model.
 Background In transcriptomic research, we often want to determine if genes are unregulated or down-regulated under a particular perturbation. For example, we have a medication that may cure type 2 diabetes. In our experiment, 6 patients are split into two groups, with 3 patients taking the medication, and 3 patients taking the placebo. The patients&#39; blood samples are then collected to measure the transcriptomic profile (mRNA abundance level for each gene) using NGS technology (RNA seq)."/>
<meta name="application-name" content="Taotao&#39;s Blog Posts">
<meta name="apple-mobile-web-app-title" content="Taotao&#39;s Blog Posts"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png"><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png"><link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="/site.webmanifest"><link rel="canonical" href="https://jasontan-code.github.io/posts/gene-exp1/" /><link rel="prev" href="https://jasontan-code.github.io/posts/calculate_svd/" /><link rel="next" href="https://jasontan-code.github.io/posts/gene-exp2/" /><link rel="stylesheet" href="/lib/normalize/normalize.min.css"><link rel="stylesheet" href="/css/style.min.css"><link rel="stylesheet" href="/lib/fontawesome-free/all.min.css"><link rel="stylesheet" href="/lib/animate/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "Model the Gene Expression (1): A GLM framework",
        "inLanguage": "en",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "https:\/\/jasontan-code.github.io\/posts\/gene-exp1\/"
        },"genre": "posts","keywords": "Genetics","wordcount":  1404 ,
        "url": "https:\/\/jasontan-code.github.io\/posts\/gene-exp1\/","datePublished": "2021-11-23T15:18:26-06:00","dateModified": "2021-11-23T15:18:26-06:00","publisher": {
            "@type": "Organization",
            "name": ""},"author": {
                "@type": "Person",
                "name": "Taotao"
            },"description": ""
    }
    </script></head>
    <body header-desktop="fixed" header-mobile="auto"><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('auto' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : 'auto' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="/" title="Taotao Tan">Taotao Tan</a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="/posts/"> Posts </a><a class="menu-item" href="/categories/"> Categories </a><a class="menu-item" href="/about/"> About </a><span class="menu-item delimiter"></span><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                    <i class="fas fa-adjust fa-fw"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="/" title="Taotao Tan">Taotao Tan</a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><a class="menu-item" href="/posts/" title="">Posts</a><a class="menu-item" href="/categories/" title="">Categories</a><a class="menu-item" href="/about/" title="">About</a><a href="javascript:void(0);" class="menu-item theme-switch" title="Switch Theme">
                <i class="fas fa-adjust fa-fw"></i>
            </a></div>
    </div>
</header>
<div class="search-dropdown desktop">
    <div id="search-dropdown-desktop"></div>
</div>
<div class="search-dropdown mobile">
    <div id="search-dropdown-mobile"></div>
</div>
<main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">Contents</h2>
            <div class="toc-content" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">Model the Gene Expression (1): A GLM framework</h1><div class="post-meta">
            <div class="post-meta-line"><span class="post-author"><a href="/" title="Author" rel=" author" class="author"><i class="fas fa-user-circle fa-fw"></i>Taotao</a></span>&nbsp;<span class="post-category">included in <a href="/categories/genetics/"><i class="far fa-folder fa-fw"></i>Genetics</a></span></div>
            <div class="post-meta-line"><i class="far fa-calendar-alt fa-fw"></i>&nbsp;<time datetime="2021-11-23">2021-11-23</time>&nbsp;<i class="fas fa-pencil-alt fa-fw"></i>&nbsp;1404 words&nbsp;
                <i class="far fa-clock fa-fw"></i>&nbsp;7 minutes&nbsp;</div>
        </div><div class="details toc" id="toc-static"  kept="true">
                <div class="details-summary toc-title">
                    <span>Contents</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#background">Background</a></li>
    <li><a href="#negative-binomial-distribution">Negative Binomial Distribution</a></li>
    <li><a href="#warm-up-fit-a-simple-model">Warm up: fit a simple model</a></li>
    <li><a href="#generalized-linear-model">Generalized Linear model</a></li>
    <li><a href="#summary">Summary</a></li>
    <li><a href="#reference">Reference</a></li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><p>Before you start reading this post, please familiarize yourself with <a href="https://jasontan-code.github.io/posts/mle/" target="_blank" rel="noopener noreffer">MLE</a> and linear model.</p>
<p> <br>
 </p>
<h2 id="background">Background</h2>
<p>In transcriptomic research, we often want to determine if genes are unregulated or down-regulated under a particular perturbation. For example, we have a medication that may cure type 2 diabetes. In our experiment, 6 patients are split into two groups, with 3 patients taking the medication, and 3 patients taking the placebo. The patients' blood samples are then collected to measure the transcriptomic profile (mRNA abundance level for each gene) using NGS technology (<a href="https://en.wikipedia.org/wiki/RNA-Seq" target="_blank" rel="noopener noreffer">RNA seq</a>). The mRNA abundance levels are quantified by the number of <a href="https://en.wikipedia.org/wiki/Read_%28biology%29" target="_blank" rel="noopener noreffer">reads</a> that were mapped to the reference genome. Finally, we use statistical tests to determine if the level of change is big enough to be considered as a DEG (differentially expressed gene).</p>
<p> </p>
<p>A typical RNA-seq experiment dataset (read counts) looks like:</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="/images/Gene_expression1/exp1.png"
        data-srcset="/images/Gene_expression1/exp1.png, /images/Gene_expression1/exp1.png 1.5x, /images/Gene_expression1/exp1.png 2x"
        data-sizes="auto"
        alt="/images/Gene_expression1/exp1.png"
        title="exp1" /></p>
<p><strong>In this post, we will only focus on one gene.</strong></p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="negative-binomial-distribution">Negative Binomial Distribution</h2>
<p>You might have noticed that in the table above, all the data is discrete counts. It is therefore inappropriate to use a normal distribution (a continuous distribution) to model the read counts. It is more common to use a <strong>Poisson distribution</strong> or a <strong>negative binomial distribution</strong> to model the gene expression in RNA-seq. For a Poisson distribution, it is assumed that the mean is equal to its variance, which isn&rsquo;t true in most cases. The most accepted distribution among bioinformatics/biostatistics community is to use a negative binomial distribution, which allows us to model the data when its variance is much greater than its mean.</p>
<p> </p>
<p>You might have seen that negative binomial distribution is formulated as:
$$
\text{P(x = k)} = {k + n -1 \choose k}\cdot p^k \cdot (1-p)^n
$$</p>
<p>This parameterization is advantageous when you want to model the number of failures before the experiment is stopped (e.g. how many times you need to flip a coin until you get 3 heads).</p>
<p> </p>
<p>However, we would like to interpret our model in a more straightforward way. An alternative parameterization of negative binomial distribution is to use a mean parameter $\mu$, and a dispersion parameter $r$:
$$
\text{P(x = k)} = \frac{\Gamma(r + k)}{k! \Gamma(r)} \cdot \Bigg( \frac{r}{r + \mu} \Bigg)^r \Bigg( \frac{\mu}{r + \mu} \Bigg)^k
$$</p>
<p>, where the dispersion parameter $r = \frac{\mu^2}{\sigma^2 - \mu}$.</p>
<p>(Notice: in DESeq2, the dispersion parameter is written as $\alpha = \frac{\sigma^2 - \mu}{\mu^2}$, which is simply the reciprocal of $r$. We will stick with $r$, since the formula looks cleaner)</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="warm-up-fit-a-simple-model">Warm up: fit a simple model</h2>
<p>Let&rsquo;s try to model the <strong>Gene1</strong> expression level of <strong>patient1, patient2, patient3</strong></p>
<table>
<thead>
<tr>
<th style="text-align:left">Patient</th>
<th style="text-align:right">Gene1 Expression</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">patient1</td>
<td style="text-align:right">1000</td>
</tr>
<tr>
<td style="text-align:left">patient2</td>
<td style="text-align:right">400</td>
</tr>
<tr>
<td style="text-align:left">patient3</td>
<td style="text-align:right">700</td>
</tr>
</tbody>
</table>
<p> </p>
<p>Technically, we should use <a href="https://jasontan-code.github.io/posts/mle/" target="_blank" rel="noopener noreffer">Maximum Likelihood Estimation</a> to estimate the parameters $r$, and $\mu$. The steps involve constructing a (log) likelihood function, taking the derivative, and finding $argmax$.</p>
<p>$$
\text{argmax}_{\mu, r} log \Bigg( P(y = 1000 |\mu, r) \cdot P(y = 400 |\mu, r) \cdot P(y = 700 |\mu, r) \Bigg)
$$</p>
<p>,where $P(y_i |\mu, r) = \frac{\Gamma(r + y_i )}{y_i ! \Gamma(r)} \cdot \Big( \frac{r}{r + \mu} \Big)^r \Big( \frac{\mu}{r + \mu} \Big)^{y_i} $</p>
<p>(Note: the math could be very messy, and it&rsquo;s impossible to find the analytical solution of the dispersion parameter $r$. We have to approximate $r$ by optimization approach like gradient descent)</p>
<p> </p>
<p>Here is the R code implementation:</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="n">data</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1000</span><span class="p">,</span> <span class="m">400</span><span class="p">,</span> <span class="m">700</span><span class="p">)</span>
<span class="n">neg_loglikelihood</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">params</span><span class="p">){</span>
  <span class="n">mu</span> <span class="o">=</span> <span class="n">params[1]</span>
  <span class="n">r</span> <span class="o">=</span> <span class="n">params[2]</span> <span class="c1"># dispersion</span>
  
  <span class="c1"># define the log likelihood of each patient</span>
  <span class="c1"># patient 1-3 received placebo, therefore mu = b0</span>
  <span class="n">loglikelihood1</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[1]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  <span class="n">loglikelihood2</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[2]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  <span class="n">loglikelihood3</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[3]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="n">mu</span><span class="p">))</span>
  

  <span class="c1"># summation of log, equivalent to log product.</span>
  <span class="c1"># optim function finds the minimum, we maximize the loglikelihood, which is equivalent to minimize neg_loglikelihood</span>
  <span class="nf">return</span><span class="p">(</span><span class="o">-</span> <span class="p">(</span><span class="n">loglikelihood1</span> <span class="o">+</span> <span class="n">loglikelihood2</span> <span class="o">+</span> <span class="n">loglikelihood3</span><span class="p">))</span>
<span class="p">}</span>

<span class="c1"># use optim to find argmin</span>
<span class="nf">optim</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> <span class="n">neg_loglikelihood</span><span class="p">)</span>
</code></pre></div><p>You should get
$$
\hat \mu = 700.444007 \\
\hat r =  7.635268
$$</p>
<p> </p>
<p>Here, I drew a 3D graph that shows you the relationship between log-likelihood, dispersion $r$ and mean $\mu$.</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="/images/Gene_expression1/loglikelihood.png"
        data-srcset="/images/Gene_expression1/loglikelihood.png, /images/Gene_expression1/loglikelihood.png 1.5x, /images/Gene_expression1/loglikelihood.png 2x"
        data-sizes="auto"
        alt="/images/Gene_expression1/loglikelihood.png"
        title="/images/Gene_expression1/loglikelihood.png" /></p>
<p>We can roughly tell that when $\hat \mu = 700$, $\hat r = 7.6$, the log likelihood function reach its maximum. These two values are so-called &ldquo;population parameters&rdquo;. Now, if we plug the estimated parameters into the formula, we have:
$$
\text{P(x = k)} = \frac{\Gamma(7.6 + k)}{k! \Gamma(7.6)} \cdot \Bigg( \frac{7.6}{7.6 + 700} \Bigg)^{7.6} \Bigg( \frac{700}{7.6 + 700} \Bigg)^k
$$</p>
<p>, and this is our fitted model.</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="generalized-linear-model">Generalized Linear model</h2>
<p>Now we need to take the effect of medication into our consideration. Keep in mind the independent variable is the medication/placebo, and the dependent variable is the expression level of Gene1.</p>
<table>
<thead>
<tr>
<th style="text-align:left">Patient</th>
<th style="text-align:right">Treatment</th>
<th style="text-align:right">Gene1 Expression</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">patient1</td>
<td style="text-align:right">placebo (0)</td>
<td style="text-align:right">1000</td>
</tr>
<tr>
<td style="text-align:left">patient2</td>
<td style="text-align:right">placebo (0)</td>
<td style="text-align:right">400</td>
</tr>
<tr>
<td style="text-align:left">patient3</td>
<td style="text-align:right">placebo (0)</td>
<td style="text-align:right">700</td>
</tr>
<tr>
<td style="text-align:left">patient4</td>
<td style="text-align:right">medication (1)</td>
<td style="text-align:right">2100</td>
</tr>
<tr>
<td style="text-align:left">patient5</td>
<td style="text-align:right">medication (1)</td>
<td style="text-align:right">3200</td>
</tr>
<tr>
<td style="text-align:left">patient6</td>
<td style="text-align:right">medication (1)</td>
<td style="text-align:right">1000</td>
</tr>
</tbody>
</table>
<p>We will treat placebo as $0$, and medication as $1$.</p>
<p>We model the gene expression as</p>
<p>$$
y \sim NB(\mu, r) \\
log( \mu )= b_1 x + b_0
$$</p>
<p>It&rsquo;s worth noticing that $b_1, b_0$, and $r$ are unknown, and the goal is to estimate them from our data.</p>
<p> </p>
<p>The formula is confusing, isn&rsquo;t it? Let&rsquo;s plug in some numbers to see if the expression makes sense.
Take patient1 as an example, we know patients 1 didn&rsquo;t receive any medication, so we have $x_1 = 0$.
$$
log(\mu_1) = b_1 \times 0 + b_0 \\
\mu_1 = e^{b_0}
$$</p>
<p>Then we plug $\mu_1$ into the formula of negative binomial distribution:
$$
NB(e^{b_0}, r )
$$</p>
<p>We observed the gene1 expression level for <strong>patient1</strong>, so we can formulate the likelihood function for patient1 as:
$$
P(y_1 = 1000 |e^{b_0}, r ) =  \frac{\Gamma(r + 1000)}{1000! \Gamma(r)} \cdot \Bigg( \frac{r}{r + e^{b_0}} \Bigg)^r \Bigg( \frac{e^{b_0}}{r + e^{b_0}} \Bigg)^{1000}
$$</p>
<p> <br>
 </p>
<p>Similarly, we can express the likelihood function of all the patients:</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="/images/Gene_expression1/likelihood2.png"
        data-srcset="/images/Gene_expression1/likelihood2.png, /images/Gene_expression1/likelihood2.png 1.5x, /images/Gene_expression1/likelihood2.png 2x"
        data-sizes="auto"
        alt="/images/Gene_expression1/likelihood2.png"
        title="/images/Gene_expression1/likelihood2.png" /></p>
<p>Since samples (patients) are independently and identically distribution, the likelihood of <strong>observing all the patients</strong> is the product of the last column of the table above.
$$
f(r, b_0, b_1) = \prod_{i = 1}^3 P(y_i |e^{b_0}, r ) \cdot \prod_{i = 4}^6 P(y_i |e^{b_0 + b1}, r )
$$</p>
<p>Notice the likelihood function $f(r, b_0, b_1)$ is a function of $r, b_0, b_1$, and we could use the same approach to find a set of parameters $\hat r, \hat b_0, \hat b_1$ that maximizes $f(r, b_0, b_1)$. Again, the actual calculation is tedious and unmanageable, and we should use <code>optim()</code> function to help us.</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="c1"># observed data</span>
<span class="n">data</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1000</span><span class="p">,</span> <span class="m">400</span><span class="p">,</span> <span class="m">700</span><span class="p">,</span> <span class="m">2100</span><span class="p">,</span> <span class="m">3200</span><span class="p">,</span> <span class="m">1000</span><span class="p">)</span>


<span class="n">neg_loglikelihood</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">params</span><span class="p">){</span>
  <span class="n">b0</span> <span class="o">=</span> <span class="n">params[1]</span> <span class="c1"># intercept</span>
  <span class="n">b1</span> <span class="o">=</span> <span class="n">params[2]</span> <span class="c1"># coefficient</span>
  <span class="n">r</span> <span class="o">=</span> <span class="n">params[3]</span> <span class="c1"># dispersion</span>
  
  <span class="c1"># define the log likelihood of each patient</span>
  <span class="c1"># patient 1-3 received placebo, therefore mu = b0</span>
  <span class="n">loglikelihood1</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[1]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span><span class="p">)))</span>
  <span class="n">loglikelihood2</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[2]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span><span class="p">)))</span>
  <span class="n">loglikelihood3</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[3]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span><span class="p">)))</span>
  
  <span class="c1"># patient 4-6 received medication, therefore mu = b0 + b1</span>
  <span class="n">loglikelihood4</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[4]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span> <span class="o">+</span> <span class="n">b1</span><span class="p">)))</span>
  <span class="n">loglikelihood5</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[5]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span> <span class="o">+</span> <span class="n">b1</span><span class="p">)))</span>
  <span class="n">loglikelihood6</span> <span class="o">=</span> <span class="nf">log</span><span class="p">(</span><span class="nf">dnbinom</span><span class="p">(</span><span class="n">data[6]</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">r</span><span class="p">,</span> <span class="n">mu</span> <span class="o">=</span> <span class="nf">exp</span><span class="p">(</span><span class="n">b0</span> <span class="o">+</span> <span class="n">b1</span><span class="p">)))</span>

  <span class="c1"># summation of log, equivalent to log product.</span>
  <span class="c1"># optim function finds the minimum, we maximize the loglikelihood, which is equivalent to minimize neg_loglikelihood</span>
  <span class="nf">return</span><span class="p">(</span><span class="o">-</span> <span class="p">(</span><span class="n">loglikelihood1</span> <span class="o">+</span> <span class="n">loglikelihood2</span> <span class="o">+</span> <span class="n">loglikelihood3</span> <span class="o">+</span> <span class="n">loglikelihood4</span> <span class="o">+</span> <span class="n">loglikelihood5</span> <span class="o">+</span> <span class="n">loglikelihood6</span><span class="p">))</span>
<span class="p">}</span>


<span class="nf">optim</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> <span class="n">neg_loglikelihood</span><span class="p">)</span>
</code></pre></div><p>The <code>optim</code> function helped us find a set of parameters that maximize the loglikelihood function: $$\hat b_0 = 6.6 \\\ \hat b_1 =1.1 \\\ \hat r = 5.9$$</p>
<p> <br>
 </p>
<p>You might want to use <code>glm.nb()</code> function from the <code>MASS</code> package to fit the model. It should give us the same results</p>
<div class="highlight"><pre class="chroma"><code class="language-r" data-lang="r"><span class="nf">library</span><span class="p">(</span><span class="n">MASS</span><span class="p">)</span>

<span class="c1"># write data</span>
<span class="n">data</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1000</span><span class="p">,</span> <span class="m">400</span><span class="p">,</span> <span class="m">700</span><span class="p">,</span> <span class="m">2100</span><span class="p">,</span> <span class="m">3200</span><span class="p">,</span> <span class="m">1000</span><span class="p">)</span>

<span class="c1"># define group</span>
<span class="n">expgroup</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;control&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span><span class="p">,</span><span class="s">&#34;treatment&#34;</span> <span class="p">))</span>

<span class="c1"># build model</span>
<span class="nf">summary.glm</span><span class="p">(</span><span class="nf">glm.nb</span><span class="p">(</span><span class="n">data</span> <span class="o">~</span> <span class="n">expgroup</span><span class="p">))</span>
</code></pre></div><p> <br>
 <br>
 <br>
 </p>
<h2 id="summary">Summary</h2>
<p>Generalized linear model is a very powerful approach that gives us more flexibility when linearity and normality don&rsquo;t hold. Besides modeling the gene expression, GLM could also be used as a classification algorithm (logistic regression), which is something I will cover in the future.</p>
<p>Maximum likelihood estimation plays a critical role in GLM. It gives us a framework of estimating the population parameters, and helps us think probabilistically. In the next post, we will use MLE to perform hypothesis testing (likelihood ratio test). Stay tuned!</p>
<p> <br>
 <br>
 <br>
 </p>
<h2 id="reference">Reference</h2>
<p>Wiki: <a href="https://en.wikipedia.org/wiki/Negative_binomial_distribution">https://en.wikipedia.org/wiki/Negative_binomial_distribution</a></p>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>Updated on 2021-11-23</span>
            </div>
            <div class="post-info-license"></div>
        </div>
        <div class="post-info-line">
            <div class="post-info-md"></div>
            <div class="post-info-share">
                <span><a href="javascript:void(0);" title="Share on Twitter" data-sharer="twitter" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework" data-via="doubleTaoTan" data-hashtags="Genetics"><i class="fab fa-twitter fa-fw"></i></a><a href="javascript:void(0);" title="Share on Facebook" data-sharer="facebook" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-hashtag="Genetics"><i class="fab fa-facebook-square fa-fw"></i></a><a href="javascript:void(0);" title="Share on WhatsApp" data-sharer="whatsapp" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework" data-web><i class="fab fa-whatsapp fa-fw"></i></a><a href="javascript:void(0);" title="Share on Line" data-sharer="line" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework"><i data-svg-src="/lib/simple-icons/icons/line.min.svg"></i></a><a href="javascript:void(0);" title="Share on 微博" data-sharer="weibo" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework"><i class="fab fa-weibo fa-fw"></i></a><a href="javascript:void(0);" title="Share on Myspace" data-sharer="myspace" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework" data-description=""><i data-svg-src="/lib/simple-icons/icons/myspace.min.svg"></i></a><a href="javascript:void(0);" title="Share on Blogger" data-sharer="blogger" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework" data-description=""><i class="fab fa-blogger fa-fw"></i></a><a href="javascript:void(0);" title="Share on Evernote" data-sharer="evernote" data-url="https://jasontan-code.github.io/posts/gene-exp1/" data-title="Model the Gene Expression (1): A GLM framework"><i class="fab fa-evernote fa-fw"></i></a></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"><i class="fas fa-tags fa-fw"></i>&nbsp;<a href="/tags/genetics/">Genetics</a></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">Back</a></span>&nbsp;|&nbsp;<span><a href="/">Home</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="/posts/calculate_svd/" class="prev" rel="prev" title="Calculate SVD by hand (and decompose Spongebob)"><i class="fas fa-angle-left fa-fw"></i>Calculate SVD by hand (and decompose Spongebob)</a>
            <a href="/posts/gene-exp2/" class="next" rel="next" title="Model the Gene Expression (2): Likelihood Ratio Test">Model the Gene Expression (2): Likelihood Ratio Test<i class="fas fa-angle-right fa-fw"></i></a></div>
</div>
<div id="comments"><div id="fb-root" class="comment"></div>
            <div
                class="fb-comments"
                data-href="https://jasontan-code.github.io/posts/gene-exp1/"
                data-width="100%"
                data-numposts="10"
            ></div><noscript>
                Please enable JavaScript to view the comments powered by <a href="https://developers.facebook.com/docs/plugins/comments/"></a>Facebook</a>.
            </noscript></div></article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">Powered by <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.80.0">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"><i class="far fa-kiss-wink-heart fa-fw"></i> LoveIt</a>
                </div><div class="footer-line"><i class="far fa-copyright fa-fw"></i><span itemprop="copyrightYear">2021</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="/" target="_blank"></a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="Back to Top">
                <i class="fas fa-arrow-up fa-fw"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="View Comments">
                <i class="fas fa-comment fa-fw"></i>
            </a>
        </div><link rel="stylesheet" href="/lib/lightgallery/lightgallery.min.css"><link rel="stylesheet" href="/lib/katex/katex.min.css"><link rel="stylesheet" href="/lib/katex/copy-tex.min.css"><link rel="stylesheet" href="/lib/cookieconsent/cookieconsent.min.css"><script type="text/javascript" src="https://connect.facebook.net/en_US/sdk.js#xfbml=1&amp;version=v5.0&amp;appId=2162149760591137&amp;autoLogAppEvents=1" defer></script><script type="text/javascript" src="/lib/smooth-scroll/smooth-scroll.min.js"></script><script type="text/javascript" src="/lib/lazysizes/lazysizes.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lightgallery.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lg-thumbnail.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lg-zoom.min.js"></script><script type="text/javascript" src="/lib/clipboard/clipboard.min.js"></script><script type="text/javascript" src="/lib/sharer/sharer.min.js"></script><script type="text/javascript" src="/lib/katex/katex.min.js"></script><script type="text/javascript" src="/lib/katex/auto-render.min.js"></script><script type="text/javascript" src="/lib/katex/copy-tex.min.js"></script><script type="text/javascript" src="/lib/katex/mhchem.min.js"></script><script type="text/javascript" src="/lib/cookieconsent/cookieconsent.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"Copy to clipboard","maxShownLines":10},"comment":{},"cookieconsent":{"content":{"dismiss":"Got it!","link":"Learn more","message":"This website uses Cookies to improve your experience."},"enable":true,"palette":{"button":{"background":"#f0f0f0"},"popup":{"background":"#1aa3ff"}},"theme":"edgeless"},"lightGallery":{"actualSize":false,"exThumbImage":"data-thumbnail","hideBarsDelay":2000,"selector":".lightgallery","speed":400,"thumbContHeight":80,"thumbWidth":80,"thumbnail":true},"math":{"delimiters":[{"display":true,"left":"$$","right":"$$"},{"display":true,"left":"\\[","right":"\\]"},{"display":false,"left":"$","right":"$"},{"display":false,"left":"\\(","right":"\\)"}],"strict":false}};</script><script type="text/javascript" src="/js/theme.min.js"></script><script type="text/javascript">
            window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments);}gtag('js', new Date());
            gtag('config', 'G-C4MZRQHWVE', { 'anonymize_ip': true });
        </script><script type="text/javascript" src="https://www.googletagmanager.com/gtag/js?id=G-C4MZRQHWVE" async></script></body>
</html>
